docs/rfc/rfc-008-don't-panic.md

```diff
@@ -0,0 +1,133 @@
+# RFC 008: Don't Panic
+
+## Changelog
+
+- 2021-12-17: initial draft (@tychoish)
+
+## Abstract
+
+Today, the Tendermint core codebase has panics in a number of cases as
+a response to exceptional situations. These panics complicate testing,
+and might make tendermint components difficult to use as a library in
+some circumstances. This document outlines a project of converting
+panics to errors and describes the situations where its safe to
+panic.
+
+## Background
+
+Panics in Go are a great mechanism for aborting the current execution
+for truly exceptional situations (e.g. memory errors, data corruption,
+processes initialization); however, because they resemble exceptions
+in other languages, it can be easy to over use them in the
+implementation of software architectures. This certainly happened in
+the history of Tendermint, and as we embark on the project of
+stabilizing the package, we find ourselves in the right moment to
+reexamine our use of panics, and largely where panics happen in the
+code base.
+
+There are still some situations where panics are acceptable and
+desireable, but it's important that Tendermint, as a project, comes to
+consensus--perhaps in the text of this document--on the situations
+where it is acceptable to panic.
+
+### References
+
+- [Defer Panic and Recover](https://go.dev/blog/defer-panic-and-recover)
+- [Why Go gets exceptions right](https://dave.cheney.net/tag/panic)
+- [Don't panic](https://dave.cheney.net/practical-go/presentations/gophercon-singapore-2019.html#_dont_panic)
+
+## Discussion
+
+### Acceptable Panics
+
+#### Initialization
+
+It is unambiguously safe (and desireable) to panic in `init()`
+functions in response to any kind of error. These errors are caught by
+tests, and occur early enough in process initialization that they
+won't cause unexpected runtime crashes.
+
+Other code that is called early in process initialization MAY panic,
+in some situations if it's not possible to return an error or cause
+the process to abort early, although these situations should be
+vanishingly slim.
+
+#### Data Corruption
+
+If Tendermint code encounters an inconsistency that could be
+attributed to data corruption or a logical impossibility it is safer
+to panic and crash the process than continue to attempt to make
+progress in these situations.
+
+Examples including reading data out of the storage engine that
+is invalid or corrupt, or encountering an ambiguous situation where
+the process should halt. Generally these forms of corruption are
```

Comment:
I'm wondering if you think that this should include potentially ambiguous cases in the consensus logic. 
An example is this line of code
https://github.com/tendermint/tendermint/blob/ab7da86b06233b325406d8b0fb5ff0902adec001/internal/consensus/state.go#L1511

We panic here because locally, we think that the block we are looking at is invalid, but for some reason a large part of the network did not think that it was invalid. We likely want to shutdown the process because it's likely that the program can no longer make meaningful progress: it can't validate future blocks if the network is behaving in a way it doesn't understand etc. Therefore, we likely want to stop, but I'm not certain if panicking makes sense in this case?

Reply:
Yes, I think the important "test" is whether anything meaningful can be done with an error.  

"Meaningful" could be as simple as logging, failing a transaction, and moving on. But if continuing would leave the system in an incoherent state (which I think is what you're describing) it would be better to terminate the program. In those cases a panic is fine, and probably preferable to an explicit exit, since it lets you shut down cleanly and also write tests that exercise the condition.

Reply:
I think this is a good place to panic based on your description of what is happening here.

Reply:
I think it'd be worth adding that somewhat explicitly to this document. Something along the lines of

```
In general, a panic should be used in the case of unrecoverable consensus failures. If a process detects that the network is behaving in an incoherent way and it does not have a clearly defined and mechanism for recovering, the process should panic.
```